Add support for rendering a custom Banner on top of Element

maheichyk opened the pull request:
This PR suggests to extend module API with possibility to provide a custom Banner component to be rendered on top of Element.

weeman1337 commented on package.json:
```diff
@@ -43,6 +43,7 @@
     "eslint-plugin-import": "^2.26.0",
     "eslint-plugin-matrix-org": "^0.4.0",
     "jest": "^27.5.1",
+    "react": "17.0.2",
```
React is used under `src` → dev dependencies is the wrong place.
I would suggest to move it to `peerDependencies`.

maheichyk replied:
changed that, thanks for notice

maheichyk replied:
If this change is applied, then `BannerLifecycle.test.ts` is failing with: Cannot find module 'react', so it is used in tests. So reverted the change.

weeman1337 replied:
Probably it should then be both: A dev and a peer dependency.

weeman1337 commented on test/lifecycles/BannerLifecycle.test.ts:
```diff
@@ -0,0 +1,49 @@
+        const opts: BannerOpts = {banner: undefined};
+        module.emit(BannerLifecycle.Banner, opts);
+
+        expect(opts.banner).toEqual(customBanner);
```
```suggestion
        expect(opts.banner).toBe(customBanner);
```

weeman1337 commented on src/lifecycles/BannerLifecycle.ts:
```diff
@@ -0,0 +1,26 @@
+}
+
+export type BannerOpts = {
+    banner: JSX.Element | undefined;
```
Here should be a comment what this is and when it is used.

maheichyk replied:
comments are added

weeman1337 commented on src/lifecycles/BannerLifecycle.ts:
```diff
@@ -0,0 +1,26 @@
+*/
+
+export enum BannerLifecycle {
+    Banner = "banner",
```
There should be a comment what this is, when it is invoked and how to use it.

maheichyk replied:
updated

weeman1337 reviewed (changes requested):
Thank you for your contribution @maheichyk .

- There should be another PR in React-SDK to actually display the banner. Can you link it here?
- The banner feature should be mentioned in the README.

maheichyk commented:
@weeman1337 thanks for your comments, I have provided the changes. Here is a PR in matrix-react-sdk https://github.com/matrix-org/matrix-react-sdk/pull/10321 that is linked. Is it fine just to keep this link just in a comment or should I move it to the description?

weeman1337 commented:
@maheichyk sorry this issue got out of sight. Usually the workflow here is to re-request a review one changes are done.

![image](https://user-images.githubusercontent.com/6216686/233289700-fb201a8c-6a9c-4898-a5ca-d5eb727a1e80.png)

This way it will be on my review list :)

> Is it fine just to keep this link just in a comment or should I move it to the description?

For a better discoverability add it to the description please.

maheichyk commented:
@germain-gg we discussed this PR last Wednesday, it seems to be okay. Could we move this forward?

germain-gg commented on package.json:
```diff
@@ -44,11 +44,15 @@
     "eslint-plugin-import": "^2.26.0",
     "eslint-plugin-matrix-org": "^0.4.0",
     "jest": "^27.5.1",
+    "react": "17.0.2",
```
```suggestion
```

maheichyk replied:
it is used by the test currently, so I think we need it.

germain-gg reviewed (changes requested):
The banner use case makes a lot of sense, but after discussing with other team members internally we believe we would like to make this module a bit more generic.

Rather than a banner, we would like to suggest defining a `wrapper`. 

Similarly to the `banner`, you should be able to define a React Component that will take on a children prop.
From an implementation point of view it would wrap the `MatrixChat` component, and let any consumer add a banner, a footer.

This brings the flexibility of a pseudo-iframe without the constraints.

maheichyk commented:
@germain-gg thanks a lot for the suggestion, sounds a good idea. I have added `WrapperLifecycle` and removed the banner related code. Please have a look.

germain-gg commented on test/lifecycles/WrapperLifecycle.test.tsx:
```diff
@@ -0,0 +1,52 @@
+
+            protected wrapperListener: WrapperListener = (wrapperOpts: WrapperOpts) => {
+                wrapperOpts.Wrapper = ({ children }) => {
+                    return <><div>Header</div>{children}<div>Footer</div></>;
```
Let's use semantic elements here

```suggestion
                    return <>
                        <header>Header</header>
                        {children}
                        <footer>Footer</footer>
                    </>;
```

maheichyk replied:
Thanks for idea

germain-gg commented on test/lifecycles/WrapperLifecycle.test.tsx:
```diff
@@ -0,0 +1,52 @@
+        module.emit(WrapperLifecycle.Wrapper, opts);
+
+        render(<opts.Wrapper><span>MatrixChat</span></opts.Wrapper>);
+        screen.getByText(/Header/i);
+        screen.getByText(/MatrixChat/i);
+        screen.getByText(/Footer/i);
```
I think we should write proper assertions rather than relying on the fact that `screen.getBy` will throw if the item is not found.

```
expect(screen.getByRole("banner")).toBeInTheDocument()
```

And we should rely on aria roles when possible. `header` uses the [`banner` role](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/banner_role) by default and `footer` uses the [`contentinfo` role](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/contentinfo_role)

We should also add an assertion to make sure that the header/footer are sibling elements to the Matrix chat.

maheichyk replied:
Thanks for the comment, sounds all good, I have updated the test

germain-gg reviewed (changes requested):
Thank you for the updates, couple of changes in the test and that should be ready to be merged.

germain-gg reviewed (approved):
Looks good to me, thank you for updating everything.
That should be good to go 🎉

germain-gg commented:
@maheichyk could you update your branch with the base branch? It's preventing me from merging that pull request otherwise. And I do not have permissions as maintainer to push to your branch.

maheichyk commented:
@germain-gg thanks for the merge! Could you please create a release of the module api, then I could finish: https://github.com/vector-im/element-web/pull/25537

germain-gg commented:
@maheichyk It's been released as part of `2.1.0`

maheichyk commented:
@fnwbr yes, it is used in our [opendesk-module](https://github.com/nordeck/element-web-modules/tree/main/packages/element-web-opendesk-module) that is public since some time now.
